File: leo2-common/src/main/java/org/deku/leo2/messaging/log/LogAppender.java
package org.deku.leo2.messaging.log;

import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.classic.spi.LoggingEvent;
import ch.qos.logback.core.AppenderBase;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.deku.leo2.messaging.MessagingContext;
import org.deku.leo2.messaging.log.v1.LogMessage;
import sx.Disposable;
import sx.jms.MessageConverter;
import sx.jms.converters.DefaultMessageConverter;
import sx.jms.embedded.Broker;

import javax.jms.Connection;
import javax.jms.DeliveryMode;
import javax.jms.MessageProducer;
import javax.jms.Session;
import java.util.ArrayList;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * Log appender sending log messages via jms
 * Created by masc on 11.06.15.
 */
public class LogAppender extends AppenderBase<ILoggingEvent> implements  Disposable {
    private Log mLog = LogFactory.getLog(this.getClass());

    private MessagingContext mMessagingContext;
    /** Message converter */
    private MessageConverter mMessageConverter;
    /** Log message buffer */
    private ArrayList<LogMessage> mLogMessageBuffer = new ArrayList<>();
    /** Flush scheduler */
    private ScheduledExecutorService mScheduledExecutorService;

    /**
     * Broker listener, jms destination is automatically created when broker start is detected
     */
    Broker.EventListener mBrokerEventListener = new Broker.EventListener() {
        @Override
        public void onStart() {
            mScheduledExecutorService.scheduleAtFixedRate(
                    () -> flush(),
                    0,
                    5,
                    TimeUnit.SECONDS);
        }

        @Override
        public void onStop() {
            dispose();
        }
    };

    /**
     * c'tor
     * @param messagingContext
     */
    public LogAppender(MessagingContext messagingContext) {
        mMessagingContext = messagingContext;

        mMessageConverter = new DefaultMessageConverter(
                DefaultMessageConverter.SerializationType.KRYO,
                DefaultMessageConverter.CompressionType.GZIP);
    }

    /**
     * Flush log messages to underlying jms broker
     */
    private void flush() {
        ArrayList<LogMessage> logMessageBuffer;

        synchronized(mLogMessageBuffer) {
            logMessageBuffer = new ArrayList(mLogMessageBuffer);
            mLogMessageBuffer.clear();
        }

        if (logMessageBuffer.size() > 0) {
            mLog.trace(String.format("Flushing [%d]", logMessageBuffer.size()));
            try {
                Connection cn = mMessagingContext.getBroker().getConnectionFactory().createConnection();
                cn.start();
                Session session = cn.createSession(true, Session.AUTO_ACKNOWLEDGE);

                MessageProducer mp = session.createProducer(mMessagingContext.getCentralLogQueue());
                mp.setDeliveryMode(DeliveryMode.PERSISTENT);
                // Log messages live a few days before they are purged by the broker
                mp.setTimeToLive(TimeUnit.DAYS.toMillis(2));
                mp.setPriority(1);
                mp.send(mMessageConverter.toMessage(
                        logMessageBuffer.toArray(new LogMessage[0]),
                        session
                ));

                session.commit();;
            } catch(Exception e) {
                mLog.error(e.getMessage(), e);
            }
        }
    }

    @Override
    protected void append(ILoggingEvent eventObject) {
        LoggingEvent le = (LoggingEvent)eventObject;
        synchronized(mLogMessageBuffer) {
            mLogMessageBuffer.add(new LogMessage(le));
        }
    }

    @Override
    public void start() {
        if (mScheduledExecutorService != null) {
            this.stop();
        }

        mScheduledExecutorService = Executors.newScheduledThreadPool(1);

        mMessagingContext.getBroker().getDelegate().add(mBrokerEventListener);
        if (mMessagingContext.getBroker().isStarted())
            mBrokerEventListener.onStart();

        super.start();
    }

    @Override
    public void stop() {
        if (mScheduledExecutorService != null) {
            // Immediate flush and subsequent shutdown
            mScheduledExecutorService.schedule(() -> flush(), 0, TimeUnit.SECONDS);
            mScheduledExecutorService.shutdown();

            // Wait for termination
            try {
                mScheduledExecutorService.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            mScheduledExecutorService = null;
        }

        super.stop();
    }

    @Override
    public void dispose() {
        this.stop();
    }
}

File: libs/leoz-common/src/main/java/org/deku/leoz/log/LogAppender.kt
package org.deku.leoz.log

import ch.qos.logback.classic.spi.ILoggingEvent
import ch.qos.logback.classic.spi.LoggingEvent
import ch.qos.logback.core.AppenderBase
import org.deku.leoz.Identity
import org.deku.leoz.config.messaging.MessagingConfiguration
import org.slf4j.LoggerFactory
import sx.Disposable
import sx.Lifecycle
import sx.jms.Channel
import sx.jms.Broker
import java.time.Duration
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService

/**
 * Log appender sending log messages via jms
 * Created by masc on 11.06.15.
 */
class LogAppender(
        /** Messaging context */
        private val broker: Broker,
        private val logChannelConfiguration: Channel.Configuration,
        private val identitySupplier: () -> Identity)
:
        AppenderBase<ILoggingEvent>(),
        Lifecycle,
        Disposable {

    private val log = LoggerFactory.getLogger(this.javaClass)
    /** Log message buffer  */
    private val buffer = ArrayList<LogMessage.LogEntry>()

    /**
     * Flush Service
     */
    inner private class Service(executorService: ScheduledExecutorService)
    :
            sx.concurrent.Service(
                    executorService = executorService,
                    period = Duration.ofSeconds(5),
                    interruptOnCancel = false) {
        /**
         * Flush service implementation
         */
        override fun run() {
            // Flush log messages to underlying jms broker
            val logMessageBuffer = ArrayList<LogMessage.LogEntry>()

            synchronized (this@LogAppender.buffer) {
                logMessageBuffer.addAll(this@LogAppender.buffer)
                this@LogAppender.buffer.clear()
            }

            if (logMessageBuffer.size > 0) {
                log.trace("Flushing [${logMessageBuffer.size}]")
                try {
                    Channel(this@LogAppender.logChannelConfiguration).use {
                        it.send(LogMessage(
                                this@LogAppender.identitySupplier().key,
                                logMessageBuffer.toTypedArray()))
                    }
                } catch (e: Exception) {
                    log.error(e.message, e)
                }
            }
        }
    }

    private val service: Service = Service(Executors.newSingleThreadScheduledExecutor())

    /**
     * Broker listener, jms destination is automatically created when broker start is detected
     */
    val brokerEventListener: Broker.EventListener = object : Broker.DefaultEventListener() {
        override fun onStart() {
            if (this@LogAppender.isStarted)
                service.start()
        }

        override fun onStop() {
            this@LogAppender.stop()
            if (service.isStarted)
                service.trigger()
            service.stop()
        }
    }

    init {
        this.broker.delegate.add(brokerEventListener)
    }

    override fun append(eventObject: ILoggingEvent) {
        val le = eventObject as LoggingEvent
        synchronized (buffer) {
            buffer.add(LogMessage.LogEntry(le))
        }
    }

    @Synchronized override fun start() {
        super.start()
        if (this.broker.isStarted)
            this.service.start()
    }

    @Synchronized override fun stop() {
        // Shutdown log flush gracefully
        this.service.stop()
        super.stop()
    }

    override fun restart() {
        this.stop()
        this.start()
    }

    override fun isRunning(): Boolean {
        return this.isStarted
    }

    override fun close() {
        this.stop()
    }
}
